File: packages/translation/src/__tests__/TranslationModule.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { RynBridge, MockTransport } from '@rynbridge/core';
import type { BridgeResponse } from '@rynbridge/core';
import { TranslationModule } from '../TranslationModule.js';

describe('TranslationModule', () => {
  let transport: MockTransport;
  let bridge: RynBridge;
  let translation: TranslationModule;

  beforeEach(() => {
    transport = new MockTransport();
    bridge = new RynBridge({ timeout: 5000 }, transport);
    translation = new TranslationModule(bridge);
  });

  function respondSuccess(payload: Record<string, unknown> = {}) {
    const sent = JSON.parse(transport.sent[transport.sent.length - 1]);
    const response: BridgeResponse = {
      id: sent.id,
      status: 'success',
      payload,
      error: null,
    };
    transport.simulateIncoming(JSON.stringify(response));
  }

  function simulateNativeEvent(module: string, action: string, payload: Record<string, unknown>) {
    const event = JSON.stringify({
      id: 'event-' + Math.random().toString(36).slice(2),
      module,
      action,
      payload,
      version: '1.0.0',
    });
    transport.simulateIncoming(event);
  }

  describe('translate', () => {
    it('sends correct module, action, and payload', async () => {
      const promise = translation.translate({ text: 'Hello', source: 'en', target: 'ko' });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('translation');
      expect(sent.action).toBe('translate');
      expect(sent.payload).toEqual({ text: 'Hello', source: 'en', target: 'ko' });

      respondSuccess({ text: '안녕하세요' });

      const result = await promise;
      expect(result).toEqual({ text: '안녕하세요' });
    });
  });

  describe('translateBatch', () => {
    it('sends correct module, action, and payload', async () => {
      const promise = translation.translateBatch({ texts: ['Hello', 'World'], source: 'en', target: 'ko' });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('translation');
      expect(sent.action).toBe('translateBatch');
      expect(sent.payload).toEqual({ texts: ['Hello', 'World'], source: 'en', target: 'ko' });

      respondSuccess({ results: ['안녕하세요', '세계'] });

      const result = await promise;
      expect(result).toEqual({ results: ['안녕하세요', '세계'] });
    });
  });

  describe('detectLanguage', () => {
    it('sends correct module, action, and payload', async () => {
      const promise = translation.detectLanguage({ text: '안녕하세요' });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('translation');
      expect(sent.action).toBe('detectLanguage');
      expect(sent.payload).toEqual({ text: '안녕하세요' });

      respondSuccess({ language: 'ko', confidence: 0.95 });

      const result = await promise;
      expect(result).toEqual({ language: 'ko', confidence: 0.95 });
    });
  });

  describe('getSupportedLanguages', () => {
    it('sends correct module and action', async () => {
      const promise = translation.getSupportedLanguages();

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('translation');
      expect(sent.action).toBe('getSupportedLanguages');
      expect(sent.payload).toEqual({});

      respondSuccess({ languages: ['en', 'ko', 'ja', 'zh'] });

      const result = await promise;
      expect(result).toEqual({ languages: ['en', 'ko', 'ja', 'zh'] });
    });
  });

  describe('downloadModel', () => {
    it('sends correct module, action, and payload', async () => {
      const promise = translation.downloadModel({ language: 'ko' });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('translation');
      expect(sent.action).toBe('downloadModel');
      expect(sent.payload).toEqual({ language: 'ko' });

      respondSuccess({ success: true });

      const result = await promise;
      expect(result).toEqual({ success: true });
    });
  });

  describe('deleteModel', () => {
    it('sends correct module, action, and payload', async () => {
      const promise = translation.deleteModel({ language: 'ko' });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('translation');
      expect(sent.action).toBe('deleteModel');
      expect(sent.payload).toEqual({ language: 'ko' });

      respondSuccess({ success: true });

      const result = await promise;
      expect(result).toEqual({ success: true });
    });
  });

  describe('getDownloadedModels', () => {
    it('sends correct module and action', async () => {
      const promise = translation.getDownloadedModels();

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('translation');
      expect(sent.action).toBe('getDownloadedModels');
      expect(sent.payload).toEqual({});

      respondSuccess({ models: ['en', 'ko'] });

      const result = await promise;
      expect(result).toEqual({ models: ['en', 'ko'] });
    });
  });

  describe('onDownloadProgress', () => {
    it('subscribes to download progress events and returns unsubscribe function', () => {
      const received: unknown[] = [];
      const unsubscribe = translation.onDownloadProgress((data) => {
        received.push(data);
      });

      simulateNativeEvent('translation', 'downloadProgress', { language: 'ko', progress: 0.5 });

      expect(received).toHaveLength(1);
      expect(received[0]).toEqual({ language: 'ko', progress: 0.5 });

      unsubscribe();
      simulateNativeEvent('translation', 'downloadProgress', { language: 'ko', progress: 0.75 });

      expect(received).toHaveLength(1);
    });
  });

  describe('error handling', () => {
    it('propagates bridge errors', async () => {
      const promise = translation.translate({ text: 'Hello', source: 'en', target: 'ko' });

      const sent = JSON.parse(transport.sent[0]);
      const response: BridgeResponse = {
        id: sent.id,
        status: 'error',
        payload: {},
        error: { code: 'UNKNOWN', message: 'Translation service unavailable' },
      };
      transport.simulateIncoming(JSON.stringify(response));

      await expect(promise).rejects.toThrow('Translation service unavailable');
    });
  });
});

File: android/translation/src/test/kotlin/io/rynbridge/translation/TranslationModuleTest.kt
package io.rynbridge.translation

import io.rynbridge.core.*
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class TranslationModuleTest {

    @Test
    fun `translate returns translated text`() = runTest {
        val provider = MockTranslationProvider()
        val module = TranslationModule(provider)
        val handler = module.actions["translate"]!!

        val result = handler(mapOf(
            "text" to BridgeValue.string("Hello"),
            "source" to BridgeValue.string("en"),
            "target" to BridgeValue.string("ko")
        ))
        assertEquals("Translated: Hello", result["text"]?.stringValue)
    }

    @Test
    fun `translateBatch returns translated texts`() = runTest {
        val provider = MockTranslationProvider()
        val module = TranslationModule(provider)
        val handler = module.actions["translateBatch"]!!

        val result = handler(mapOf(
            "texts" to BridgeValue.array(listOf(BridgeValue.string("Hello"), BridgeValue.string("World"))),
            "source" to BridgeValue.string("en"),
            "target" to BridgeValue.string("ko")
        ))
        val results = result["results"]?.arrayValue
        assertNotNull(results)
        assertEquals(2, results!!.size)
        assertEquals("Translated: Hello", results[0].stringValue)
        assertEquals("Translated: World", results[1].stringValue)
    }

    @Test
    fun `detectLanguage returns language and confidence`() = runTest {
        val provider = MockTranslationProvider()
        val module = TranslationModule(provider)
        val handler = module.actions["detectLanguage"]!!

        val result = handler(mapOf("text" to BridgeValue.string("Hello")))
        assertEquals("en", result["language"]?.stringValue)
        assertEquals(0.95, result["confidence"]?.doubleValue)
    }

    @Test
    fun `getSupportedLanguages returns language list`() = runTest {
        val provider = MockTranslationProvider()
        val module = TranslationModule(provider)
        val handler = module.actions["getSupportedLanguages"]!!

        val result = handler(emptyMap())
        val languages = result["languages"]?.arrayValue
        assertNotNull(languages)
        assertEquals(3, languages!!.size)
        assertEquals("en", languages[0].stringValue)
    }

    @Test
    fun `downloadModel returns success`() = runTest {
        val provider = MockTranslationProvider()
        val module = TranslationModule(provider)
        val handler = module.actions["downloadModel"]!!

        val result = handler(mapOf("language" to BridgeValue.string("ko")))
        assertEquals(true, result["success"]?.boolValue)
    }

    @Test
    fun `deleteModel returns success`() = runTest {
        val provider = MockTranslationProvider()
        val module = TranslationModule(provider)
        val handler = module.actions["deleteModel"]!!

        val result = handler(mapOf("language" to BridgeValue.string("ko")))
        assertEquals(true, result["success"]?.boolValue)
    }

    @Test
    fun `getDownloadedModels returns model list`() = runTest {
        val provider = MockTranslationProvider()
        val module = TranslationModule(provider)
        val handler = module.actions["getDownloadedModels"]!!

        val result = handler(emptyMap())
        val models = result["models"]?.arrayValue
        assertNotNull(models)
        assertEquals(2, models!!.size)
        assertEquals("en", models[0].stringValue)
        assertEquals("ko", models[1].stringValue)
    }

    @Test
    fun `module name and version`() {
        val provider = MockTranslationProvider()
        val module = TranslationModule(provider)
        assertEquals("translation", module.name)
        assertEquals("0.1.0", module.version)
    }
}

private class MockTranslationProvider : TranslationProvider {
    override suspend fun translate(text: String, source: String, target: String): TranslateResult =
        TranslateResult(text = "Translated: $text")

    override suspend fun translateBatch(texts: List<String>, source: String, target: String): TranslateBatchResult =
        TranslateBatchResult(results = texts.map { "Translated: $it" })

    override suspend fun detectLanguage(text: String): DetectLanguageResult =
        DetectLanguageResult(language = "en", confidence = 0.95)

    override suspend fun getSupportedLanguages(): GetSupportedLanguagesResult =
        GetSupportedLanguagesResult(languages = listOf("en", "ko", "ja"))

    override suspend fun downloadModel(language: String): DownloadModelResult =
        DownloadModelResult(success = true)

    override suspend fun deleteModel(language: String): DeleteModelResult =
        DeleteModelResult(success = true)

    override suspend fun getDownloadedModels(): GetDownloadedModelsResult =
        GetDownloadedModelsResult(models = listOf("en", "ko"))
}
